sphinx/domains/math.py

```diff
@@ -0,0 +1,138 @@
+# -*- coding: utf-8 -*-
+"""
+    sphinx.domains.math
+    ~~~~~~~~~~~~~~~~~~~
+
+    The math domain.
+
+    :copyright: Copyright 2007-2018 by the Sphinx team, see AUTHORS.
+    :license: BSD, see LICENSE for details.
+"""
+
+from docutils import nodes
+from docutils.nodes import make_id
+
+from sphinx.domains import Domain
+from sphinx.addnodes import math_block as displaymath, math_reference
+from sphinx.locale import __
+from sphinx.roles import XRefRole
+from sphinx.util import logging
+from sphinx.util.nodes import make_refnode
+
+if False:
+    # For type annotation
+    from typing import Any, Callable, Dict, Iterable, List, Tuple  # NOQA
+    from sphinx.application import Sphinx  # NOQA
+    from sphinx.builders import Builder  # NOQA
+    from sphinx.environment import BuildEnvironment  # NOQA
+
+logger = logging.getLogger(__name__)
+
+
+class MathReferenceRole(XRefRole):
+    def result_nodes(self, document, env, node, is_ref):
+        # type: (nodes.Node, BuildEnvironment, nodes.Node, bool) -> Tuple[List[nodes.Node], List[nodes.Node]]  # NOQA
+        node['refdomain'] = 'math'
+        return [node], []
+
+
+class MathDomain(Domain):
+    """Mathematics domain."""
+    name = 'math'
+    label = 'mathematics'
+
+    initial_data = {
+        'objects': {},  # labelid -> (docname, eqno)
+    }  # type: Dict[unicode, Dict[unicode, Tuple[unicode, int]]]
+    dangling_warnings = {
+        'eq': 'equation not found: %(target)s',
+    }
+    enumerable_nodes = {  # node_class -> (figtype, title_getter)
+        displaymath: ('displaymath', None),
+        nodes.math_block: ('displaymath', None),
+    }  # type: Dict[nodes.Node, Tuple[unicode, Callable]]
+
+    def clear_doc(self, docname):
+        # type: (unicode) -> None
+        for equation_id, (doc, eqno) in list(self.data['objects'].items()):
+            if doc == docname:
+                del self.data['objects'][equation_id]
+
+    def merge_domaindata(self, docnames, otherdata):
+        # type: (Iterable[unicode], Dict) -> None
+        for labelid, (doc, eqno) in otherdata['objects'].items():
+            if doc in docnames:
+                self.data['objects'][labelid] = (doc, eqno)
+
+    def resolve_xref(self, env, fromdocname, builder, typ, target, node, contnode):
+        # type: (BuildEnvironment, unicode, Builder, unicode, unicode, nodes.Node, nodes.Node) -> nodes.Node  # NOQA
+        assert typ == 'eq'
+        docname, number = self.data['objects'].get(target, (None, None))
+        if docname:
+            if builder.name == 'latex':
```

Comment:
It would be nice if we can move this to LaTeX Transforms. MathDomain should not consider the behavior of each builder.